<!DOCTYPE html>



  


<html class="theme-next muse use-motion" lang="">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">



  
  
    
    
  <script src="/blog/lib/pace/pace.min.js?v=1.0.2"></script>
  <link href="/blog/lib/pace/pace-theme-minimal.min.css?v=1.0.2" rel="stylesheet">







<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="google5b248f7b86cbcee5.html" />














  
  
  <link href="/blog/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/blog/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/blog/css/main.css?v=5.1.4" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/blog/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/blog/images/favicon.ico?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/blog/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/blog/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="machine learning,python,bioinformatics,teaching,TA," />





  <link rel="alternate" href="/blog/atom.xml" title="WonderLand" type="application/atom+xml" />






<meta name="description" content="This is one of the two course quizzes instruction of Bioinformatics basic course in Tsinghua University. You may also find it here. The related jupyter file">
<meta name="keywords" content="machine learning,python,bioinformatics,teaching,TA">
<meta property="og:type" content="article">
<meta property="og:title" content="QUIZ Predicting maize traits using SNP data">
<meta property="og:url" content="https://www.cmwonderland.com/blog/2018/10/06/31_quiz_emaize_tutorial/index.html">
<meta property="og:site_name" content="WonderLand">
<meta property="og:description" content="This is one of the two course quizzes instruction of Bioinformatics basic course in Tsinghua University. You may also find it here. The related jupyter file">
<meta property="og:locale" content="default">
<meta property="og:image" content="http://i1.fuimg.com/640680/4d02e8859cdfd53f.png">
<meta property="og:image" content="http://i1.fuimg.com/640680/d2f0f8536f919855.png">
<meta property="og:image" content="http://i1.fuimg.com/640680/8ad266b66785548e.png">
<meta property="og:image" content="http://i1.fuimg.com/640680/870c6a096a867529.png">
<meta property="og:image" content="http://i1.fuimg.com/640680/ef4e5c183a271656.png">
<meta property="og:updated_time" content="2019-04-23T13:07:13.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="QUIZ Predicting maize traits using SNP data">
<meta name="twitter:description" content="This is one of the two course quizzes instruction of Bioinformatics basic course in Tsinghua University. You may also find it here. The related jupyter file">
<meta name="twitter:image" content="http://i1.fuimg.com/640680/4d02e8859cdfd53f.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/blog/',
    scheme: 'Muse',
    version: '5.1.4',
    sidebar: {"position":"left","display":"always","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="https://www.cmwonderland.com/blog/2018/10/06/31_quiz_emaize_tutorial/"/>





  <title>QUIZ Predicting maize traits using SNP data | WonderLand</title>
  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?6656499bfc0e07b4e20ee4975eb85f31";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <a href="https://github.com/james20141606"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_white_ffffff.png" alt="Fork me on GitHub"></a>
    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/blog/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">WonderLand</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">Somnium & Somniator</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/blog/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/blog/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/blog/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/blog/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/blog/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://www.cmwonderland.com/blog/blog/2018/10/06/31_quiz_emaize_tutorial/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="James Chen">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/blog/images/avatar.png">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="WonderLand">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">QUIZ Predicting maize traits using SNP data</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2018-10-06T20:03:19-04:00">
                2018-10-06
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/blog/categories/school-work/" itemprop="url" rel="index">
                    <span itemprop="name">school work</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/blog/2018/10/06/31_quiz_emaize_tutorial/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count disqus-comment-count"
                        data-disqus-identifier="2018/10/06/31_quiz_emaize_tutorial/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
             <span id="/blog/2018/10/06/31_quiz_emaize_tutorial/" class="leancloud_visitors" data-flag-title="QUIZ Predicting maize traits using SNP data">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Visitors&#58;</span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article&#58;</span>
                
                <span title="Words count in article">
                  3,230
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time">
                  14
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>This is one of the two course quizzes instruction of <a href="https://lulab.gitbooks.io/teaching/content/" target="_blank" rel="noopener">Bioinformatics basic course in Tsinghua University</a>. You may also find it <a href="https://lulab.gitbooks.io/teaching/content/quiz/quiz_exrna/quiz_emaize_tutorial.html" target="_blank" rel="noopener">here</a>.</p>
<p>The related <a href="https://github.com/lulab/teaching_book/blob/master/quiz/quiz_emaize/quiz_emaize_tutorial.ipynb" target="_blank" rel="noopener">jupyter file</a></p>
<a id="more"></a>
<h1 id="eMaize玉米育种挑战赛"><a href="#eMaize玉米育种挑战赛" class="headerlink" title="eMaize玉米育种挑战赛"></a>eMaize玉米育种挑战赛</h1><p>请在<a href="https://cloud.tsinghua.edu.cn/f/3f4fc999720d45f198ca/" target="_blank" rel="noopener">quiz_emaize_tutorial_shared</a>下载相关数据，并下载该<a href="https://cloud.tsinghua.edu.cn/f/def95f1d1beb4031bf2f/" target="_blank" rel="noopener">文件夹</a>下的内容，打开<code>quiz_emaize_tutorial.ipynb</code>文件阅读详细的<strong>Quiz指南</strong>。</p>
<h2 id="eMaize背景简介"><a href="#eMaize背景简介" class="headerlink" title="eMaize背景简介"></a>eMaize背景简介</h2><p>eMaize挑战赛是<a href="http://emaize.imaze.org/emaize/emaize_cn.php" target="_blank" rel="noopener">一个通过机器学习方法预测玉米性状的比赛</a>，要求我们以SNP作为特征，通过训练一个模型，<strong>对玉米的三个性状进行预测</strong>。</p>
<p>本教程将会包括：</p>
<ul>
<li>介绍数据的情况，使用方式</li>
<li>具体任务要求</li>
<li>一些机器学习的概念方法以及工具使用</li>
</ul>
<h2 id="编程工具介绍"><a href="#编程工具介绍" class="headerlink" title="编程工具介绍"></a>编程工具介绍</h2><p>大作业需要使用python完成，推荐读者使用python3。我们需要一些python的工具包来实现部分功能。推荐使用包管理软件Anaconda来预装一些必需的包以及安装其他需要的包。另外强烈建议使用jupyter notebook进行代码编辑、运行和调试。具体使用方法请参考教程<a href="https://lulab.gitbooks.io/teaching/content/part-iii.-machine-learning-basics/python_tutorial.html" target="_blank" rel="noopener">Anaconda 和 jupyter</a>相关指南。<br>如果本地缺少下列可能需要的包，请使用<code>pip</code>或者<code>conda</code>进行安装。如:</p>
<figure class="highlight cmake"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pip <span class="keyword">install</span> tqdm</span><br><span class="line">conda <span class="keyword">install</span> sklearn</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#导入必需的库</span></span><br><span class="line">%pylab inline</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">from</span> sklearn.random_projection <span class="keyword">import</span> SparseRandomProjection</span><br><span class="line"><span class="keyword">import</span> scipy.stats</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> sklearn</span><br><span class="line"><span class="keyword">import</span> h5py</span><br><span class="line"><span class="keyword">from</span> sklearn.metrics <span class="keyword">import</span> r2_score</span><br><span class="line"><span class="keyword">from</span> scipy.stats.stats <span class="keyword">import</span> pearsonr</span><br><span class="line"><span class="keyword">from</span> sklearn.linear_model <span class="keyword">import</span> LinearRegression</span><br><span class="line"><span class="keyword">import</span> xgboost</span><br><span class="line"><span class="keyword">from</span> xgboost.sklearn <span class="keyword">import</span> XGBRegressor</span><br><span class="line"><span class="keyword">from</span> sklearn.linear_model <span class="keyword">import</span> Ridge</span><br><span class="line"><span class="keyword">from</span> sklearn.kernel_ridge <span class="keyword">import</span> KernelRidge</span><br><span class="line"><span class="keyword">from</span> sklearn <span class="keyword">import</span> neighbors</span><br><span class="line"><span class="keyword">from</span> sklearn.ensemble <span class="keyword">import</span> RandomForestRegressor</span><br><span class="line"><span class="keyword">from</span> sklearn.gaussian_process <span class="keyword">import</span> GaussianProcessRegressor</span><br><span class="line"><span class="keyword">from</span> sklearn.gaussian_process.kernels <span class="keyword">import</span> DotProduct</span><br><span class="line"><span class="keyword">from</span> tqdm <span class="keyword">import</span> tqdm_notebook <span class="keyword">as</span> tqdm</span><br><span class="line"><span class="keyword">from</span> IPython.display <span class="keyword">import</span> display, Image</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">import</span> warnings</span><br><span class="line">warnings.filterwarnings(<span class="string">'ignore'</span>)</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">styles = [<span class="string">"white"</span>,<span class="string">"dark"</span>,<span class="string">'whitegrid'</span>,<span class="string">"darkgrid"</span>]</span><br><span class="line">contexts = [<span class="string">'paper'</span>,<span class="string">'talk'</span>,<span class="string">'poster'</span>,<span class="string">'notebook'</span>]</span><br><span class="line">sns.set_context(contexts[<span class="number">1</span>])</span><br><span class="line">sns.set_style(styles[<span class="number">2</span>])</span><br></pre></td></tr></table></figure>
<h2 id="数据介绍"><a href="#数据介绍" class="headerlink" title="数据介绍"></a>数据介绍</h2><p>原始数据中有6210个样本，其中4754个样本作为训练集，1456个样本作为测试集 <br></p>
<ul>
<li><strong>genotype</strong>：SNP (Single-nucleotide polymorphism) 数据<br>  每个位点可能有三种情况，如AA，AT，TT，每个样本SNP位点约为190万个，数据位置<code>data/genotype_2bit/</code>，包含十个染色体各自的SNP，使用时需要整合在一起。 <br></li>
<li><strong>trait</strong>：<br>  共三种，trait1开花期，trait2株高，trait3产量，为连续值，只提供训练集样本的性状数据。<code>data/pheno_emaize.txt</code> <br><br><br><br>作为示例仅仅使用每个样本的5000个SNP，在做大作业的过程中请使用全部的SNP。<br></li>
</ul>
<h3 id="Genotype数据"><a href="#Genotype数据" class="headerlink" title="Genotype数据"></a>Genotype数据</h3><h4 id="SNP数据存储格式"><a href="#SNP数据存储格式" class="headerlink" title="SNP数据存储格式"></a>SNP数据存储格式</h4><p>txt存储格式不适合大数据读取的问题，对内存的占用过多 <br><br>对于结构化的、能够存储为矩阵的数据，可以使用HDF5格式存取，内存占用小，读取速度快。</p>
<blockquote>
<p>tips: 在命令行查看数据shape的方法为：<br>cd至文件路径下，输入：h5ls filename</p>
</blockquote>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#使用h5py读取5000个SNP，此时数据为原始的碱基信息</span></span><br><span class="line"><span class="keyword">with</span> h5py.File(<span class="string">'data/snp_5000'</span>,<span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">    <span class="keyword">print</span> (list(f.keys()))</span><br><span class="line">    snps = f[<span class="string">'snp'</span>][:]</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#使用h5py读取5000个SNP，此时为原始的碱基信息转化为数值信息后的数据</span></span><br><span class="line"><span class="keyword">with</span> h5py.File(<span class="string">'data/2bit_geno'</span>,<span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">    <span class="keyword">print</span> (list(f.keys()))</span><br><span class="line">    snps_2bit = f[<span class="string">'data'</span>][:]</span><br><span class="line">snps_2bit.shape</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#使用h5py读取一个染色体的SNP示例，此时文件内为我们处理好的数值化数据，注意需要先解压文件：</span></span><br><span class="line"><span class="keyword">with</span> h5py.File(<span class="string">'data/genotype_2bit/chr1.h5'</span>,<span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">    <span class="keyword">print</span> (list(f.keys()))</span><br><span class="line">    chr1_snps = f[<span class="string">'data'</span>][:]</span><br><span class="line">chr1_snps.shape, chr1_snps.dtype</span><br></pre></td></tr></table></figure>
<h4 id="性状数据"><a href="#性状数据" class="headerlink" title="性状数据"></a>性状数据</h4><p>使用numpy/pandas均可读取性状数据，计算时一般使用numpy.array的形式</p>
<p>前4764个样本的性状是已知的，后1454个样本性状待预测</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">traits = pd.read_csv(<span class="string">'data/pheno_emaize.txt'</span>,<span class="string">'\t'</span>)</span><br><span class="line">display(traits.head())</span><br><span class="line">display(traits.tail())</span><br></pre></td></tr></table></figure>
<h5 id="查看性状的分布情况"><a href="#查看性状的分布情况" class="headerlink" title="查看性状的分布情况"></a>查看性状的分布情况</h5><p>注意，对性状数据已经做过了normalization</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">trait1 = np.array(traits[<span class="string">'trait1'</span>])[:<span class="number">4754</span>]</span><br><span class="line">trait2 = np.array(traits[<span class="string">'trait2'</span>])[:<span class="number">4754</span>]</span><br><span class="line">trait3 = np.array(traits[<span class="string">'trait3'</span>])[:<span class="number">4754</span>]</span><br><span class="line">fig, ax = plt.subplots(<span class="number">1</span>,<span class="number">3</span>, figsize=(<span class="number">15</span>,<span class="number">4</span>))</span><br><span class="line">ax[<span class="number">0</span>].hist(trait1,bins = <span class="number">20</span>,color=<span class="string">'b'</span>,alpha=<span class="number">0.6</span>)</span><br><span class="line">ax[<span class="number">0</span>].set_title(<span class="string">'Normalized Trait1: Flowering Time'</span>,fontsize=<span class="number">14</span>)</span><br><span class="line">ax[<span class="number">1</span>].hist(trait2,bins = <span class="number">20</span>,color=<span class="string">'g'</span>,alpha=<span class="number">0.6</span>)</span><br><span class="line">ax[<span class="number">1</span>].set_title(<span class="string">'Normalized Trait2: Height'</span>,fontsize=<span class="number">14</span>)               </span><br><span class="line">ax[<span class="number">2</span>].hist(trait3,bins = <span class="number">20</span>,color=<span class="string">'r'</span>,alpha=<span class="number">0.6</span>)</span><br><span class="line">ax[<span class="number">2</span>].set_title(<span class="string">'Normalized Trait3: Yield'</span>,fontsize=<span class="number">14</span>)</span><br></pre></td></tr></table></figure>
<p><img src="http://i1.fuimg.com/640680/4d02e8859cdfd53f.png" alt="Markdown"></p>
<h2 id="Quiz具体要求"><a href="#Quiz具体要求" class="headerlink" title="Quiz具体要求"></a>Quiz具体要求</h2><p>之前的部分我们介绍了基本的背景知识，接下来我们会提出解答本题目的具体要求：</p>
<ul>
<li>完成<strong>特征选择和特征筛除工作</strong>。</li>
<li>完成<strong>对三种性状的预测</strong>并提交预测结果，允许多次提交预测结果以获得更好的结果。</li>
<li>提交一份<strong>工作报告</strong>，中英文不限，同时提交<strong>源代码</strong>。</li>
<li>选择性完成加分项内容。</li>
</ul>
<h3 id="特征选择和冗余特征筛除"><a href="#特征选择和冗余特征筛除" class="headerlink" title="特征选择和冗余特征筛除"></a>特征选择和冗余特征筛除</h3><p>本挑战原始特征数量接近2,000,000，超过大多数机器学习模型的输入限制，特征间相关性很强，冗余特征很多，且过多的特征数量导致计算开销非常大，这都需要完成特征选择和去除冗余的步骤。</p>
<p>鉴于本问题原始特征数量过于巨大，并不是每种特征选择和降维方法都适合，请读者思考和选择合适的特征选择与降维方法，在这里仅提供几个方法参考：</p>
<ul>
<li>特征选择：<ul>
<li>ANOVA（方差分析）</li>
<li>基于模型权重排序（如线性模型）</li>
</ul>
</li>
<li>降维：<ul>
<li>PCA</li>
<li>SVD</li>
<li>Random Projection</li>
</ul>
</li>
</ul>
<blockquote>
<p>tips: 请思考是否需要针对特征选择或者降维后的数据做scale</p>
</blockquote>
<h3 id="完成对测试集三种性状的预测，尝试得到尽可能好的预测结果"><a href="#完成对测试集三种性状的预测，尝试得到尽可能好的预测结果" class="headerlink" title="完成对测试集三种性状的预测，尝试得到尽可能好的预测结果"></a>完成对测试集三种性状的预测，尝试得到尽可能好的预测结果</h3><ul>
<li>请思考和探索使用何种回归模型，读者可以尝试多种模型并比较其结果，在<em>编程工具介绍</em>部分读者可以看到一些机器学习模型的方法，也推荐读者思考和使用其他模型。</li>
<li>思考和探索是否对每个性状使用不同的模型</li>
<li>根据训练集与测试集的特殊划分方式（见<em>查看训练集与测试集的划分</em>部分）思考可以使用的策略。</li>
</ul>
<h3 id="加分内容"><a href="#加分内容" class="headerlink" title="加分内容"></a>加分内容</h3><p>为了更完整地展示emaize挑战的困难与有趣之处，我们为有余力的读者设置了更多的挑战。</p>
<h4 id="对模型进行鲁棒性测试-10’"><a href="#对模型进行鲁棒性测试-10’" class="headerlink" title="对模型进行鲁棒性测试 (10’)"></a>对模型进行鲁棒性测试 (10’)</h4><p>这是一个并不非常困难的但是对于本问题比较重要的工作，该工作可以细分为以下几项内容：</p>
<ul>
<li>在读者已有的数据（训练集数据）上进行多轮（如100、1000轮）交叉验证，测试模型的鲁棒性</li>
<li>设计不同的数据集划分方式，除了随机划分训练集与验证集，还可以有其他特殊的设计方法</li>
<li>统计测试结果，以多种形式展现，包括统计数据和图示。</li>
</ul>
<p>为了帮助读者完成测试，我们会给读者提供部分代码和一些相关的统计图，供读者使用和仿照设计，详见<em>补充知识</em>的<em>模型鲁棒性测试</em>部分。</p>
<h4 id="使用ANOVA进行特征选择的加速算法-5’"><a href="#使用ANOVA进行特征选择的加速算法-5’" class="headerlink" title="使用ANOVA进行特征选择的加速算法 (5’)"></a>使用ANOVA进行特征选择的加速算法 (5’)</h4><p>方差分析方法可以利用p值挑选feature <br><br>调用scipy.stats.f_oneway,利用SNPs和性状可以很容易地计算出p-value，进而挑选特征，但是对于大量数据来说速度较慢 <br><br>我们可以设计一种加速ANOVA计算的方法完成计算，相比于scipy.stats的方法可以提升计算速度数百倍。为了帮助读者实现这一功能，我们提供给读者设计的基本思路，请参考<a href="#fastanova"><em>ANOVA加速算法部分</em></a>，有能力的读者可以根据基本思路实现ANOVA的加速算法。</p>
<h4 id="混合线性模型-20’"><a href="#混合线性模型-20’" class="headerlink" title="混合线性模型 (20’)"></a>混合线性模型 (20’)</h4><p>育种领域的一个经典模型是混合线性模型(linear mixed model)，请尝试设计一个混合线性模型来解决本问题。</p>
<ul>
<li><p>可以研究并调用<a href="https://github.com/MicrosoftGenomics/FaST-LMM" target="_blank" rel="noopener">FaST-LMM package</a>，研究其原理并应用于我们的数据中，试图在测试集上获得好的预测结果(5’)</p>
</li>
<li><p>根据其思路进行改进，设计一个类似的混合线性模型，并且可以通过快速的交叉验证挑选超参数，最终在测试集上获得好的效果。详细内容可以参考<a href="assets/Mixed_ridge&amp;Fast_cv.pdf">这篇文档</a>。(15’)</p>
</li>
</ul>
<h2 id="补充知识（选读）"><a href="#补充知识（选读）" class="headerlink" title="补充知识（选读）"></a>补充知识（选读）</h2><h3 id="测试预测结果"><a href="#测试预测结果" class="headerlink" title="测试预测结果"></a>测试预测结果</h3><h5 id="查看训练集与测试集的划分"><a href="#查看训练集与测试集的划分" class="headerlink" title="查看训练集与测试集的划分"></a>查看训练集与测试集的划分</h5><p>请读者特别关注这个信息，训练集与测试集特殊的划分方式是本问题的一个特点与难点，也是想解决好本问题的关键。</p>
<p>下图中彩色部分为训练集性状，白色部分为待预测性状 <br><br>可以发现其划分方式并不随机，这会导致常规的机器学习方法出现一些问题，常规的机器学习问题中，每个样本彼此独立，但是对于育种问题，很多样本可能有同一个父本或者母本，比如图中的每一行样本都来自同一个父本，每一列样本都来自同一个母本。读者需要自己思考待预测样本和已有样本的关系，结合机器学习的特点“通过学习已有数据的特征对未知数据进行预测”，思考并观察预测结果的好坏。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">generate_parent_table</span><span class="params">(phenotype_file)</span>:</span></span><br><span class="line">    phenotypes = pd.read_table(phenotype_file)</span><br><span class="line">    pedigree = phenotypes[<span class="string">'pedigree'</span>].str.split(<span class="string">'_'</span>, expand=<span class="keyword">True</span>)</span><br><span class="line">    pedigree.columns = [<span class="string">'f'</span>, <span class="string">'X'</span>, <span class="string">'m'</span>]</span><br><span class="line">    phenotypes = pd.concat([phenotypes, pedigree], axis=<span class="number">1</span>)</span><br><span class="line">    phenotypes[<span class="string">'number'</span>] = np.arange(phenotypes.shape[<span class="number">0</span>])</span><br><span class="line">    parent_table = phenotypes.pivot_table(values=<span class="string">'number'</span></span><br><span class="line">                                          , index=[<span class="string">'m'</span>], columns=[<span class="string">'f'</span>], dropna=<span class="keyword">False</span>)</span><br><span class="line">    male_ids = [<span class="string">'m%d'</span> % i <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, parent_table.shape[<span class="number">0</span>] + <span class="number">1</span>)]</span><br><span class="line">    female_ids = [<span class="string">'f%d'</span> % i <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, parent_table.shape[<span class="number">1</span>] + <span class="number">1</span>)]</span><br><span class="line">    parent_table = parent_table.loc[male_ids, female_ids]</span><br><span class="line">    <span class="keyword">return</span> parent_table</span><br><span class="line">phenotype_file = <span class="string">'data/pheno_emaize.txt'</span></span><br><span class="line">parent_table = generate_parent_table(phenotype_file)</span><br><span class="line">phenotypes = pd.read_table(<span class="string">'data/pheno_emaize.txt'</span>)</span><br><span class="line">fig, ax = subplots(<span class="number">3</span>,<span class="number">1</span>, figsize=(<span class="number">15</span>, <span class="number">10</span>))</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">3</span>):</span><br><span class="line">    trait = [<span class="string">'trait1'</span>,<span class="string">'trait2'</span>,<span class="string">'trait3'</span>][i]</span><br><span class="line">    ax[i].imshow(np.take(np.ravel(phenotypes[trait].values), parent_table), cmap=cm.RdBu)</span><br><span class="line">    ax[i].set_title(<span class="string">'Phenotypes of training data (%s)'</span>%trait)</span><br><span class="line">    ax[i].set_ylabel(<span class="string">'male'</span>)</span><br><span class="line">    ax[i].set_xlabel(<span class="string">'female'</span>)</span><br><span class="line">plt.tight_layout()</span><br></pre></td></tr></table></figure>
<p><img src="http://i1.fuimg.com/640680/d2f0f8536f919855.png" alt="Markdown"></p>
<h5 id="测试集具体划分"><a href="#测试集具体划分" class="headerlink" title="测试集具体划分"></a>测试集具体划分</h5><p>我们会把测试区域分为三个部分，来测试提交的结果，三个区域分别为下图的淡蓝色，黄色和红色区域，请读者思考它们各自的特征。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">fig,ax=plt.subplots(figsize=(<span class="number">20</span>,<span class="number">5</span>))</span><br><span class="line">testexample = np.zeros([<span class="number">30</span>,<span class="number">207</span>])</span><br><span class="line">testexample[<span class="number">25</span>:,:<span class="number">202</span>] = <span class="number">100</span></span><br><span class="line">testexample[:<span class="number">25</span>,<span class="number">202</span>:] = <span class="number">200</span></span><br><span class="line">testexample[<span class="number">25</span>:,<span class="number">202</span>:] = <span class="number">300</span></span><br><span class="line">ax.imshow(testexample,cmap=<span class="string">'jet'</span>)</span><br></pre></td></tr></table></figure>
<p><img src="http://i1.fuimg.com/640680/8ad266b66785548e.png" alt="Markdown"></p>
<h5 id="评价指标"><a href="#评价指标" class="headerlink" title="评价指标"></a>评价指标</h5><p>对于回归问题，我们一般使用<script type="math/tex">r^2</script>和pearson correlation coefficient(PCC)衡量，其定义如下：</p>
<p><script type="math/tex">r^2 = 1-\frac{SS_{res}}{SS_{tot}}</script><br></p>
<p><script type="math/tex">PCC = \frac{cov(X,Y)}{\sigma_X \sigma_Y} = \frac{E[(X-\mu_X)(Y-\mu_Y)]}{\sigma_X \sigma_Y}</script><br></p>
<h3 id="将SNP数据编码为向量"><a href="#将SNP数据编码为向量" class="headerlink" title="将SNP数据编码为向量"></a>将SNP数据编码为向量</h3><p>每个位点的碱基只有三种情况，不会出现更多碱基组合的可能，比如某位点仅有AA，AT，TT三种可能的情况<br><br>我们可以采取三种方式对其编码：</p>
<ul>
<li>转化为0、1、2。找到minor allele frequency（MAF），即两种碱基（如A、T）中出现频率低的那个，以A作为MAF为例，则TT为0，AT为1，AA为2，这样可以突出MAF</li>
<li>转化为3-bit one hot vector,<script type="math/tex">[1,0,0]^T,[0,1,0]^T,[0,0,1]^T</script>这样可以保持三种向量在空间距离的一致</li>
<li>转化为2-bit vector,则AA，AT，TT分别编为<script type="math/tex">[1,0]^T,[1,1]^T,[0,1]^T</script>,不需要考虑MAF<br>我们采取第三种方式处理了数据并提供给读者</li>
</ul>
<h4 id="具体处理过程"><a href="#具体处理过程" class="headerlink" title="具体处理过程"></a>具体处理过程</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">convert_2bit</span><span class="params">(seq)</span>:</span></span><br><span class="line">    genotypes = np.zeros([<span class="number">6210</span>,<span class="number">2</span>])</span><br><span class="line">    a = seq[<span class="number">1</span>].split(<span class="string">'/'</span>)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">6210</span>):</span><br><span class="line">        <span class="keyword">if</span> seq[<span class="number">4</span>:][i] == a[<span class="number">0</span>] + a[<span class="number">0</span>]:</span><br><span class="line">            genotypes[i] = np.array([<span class="number">0</span>,<span class="number">1</span>])</span><br><span class="line">        <span class="keyword">if</span> seq[<span class="number">4</span>:][i] == a[<span class="number">0</span>] + a[<span class="number">1</span>]:</span><br><span class="line">            genotypes[i] = np.array([<span class="number">1</span>,<span class="number">0</span>])</span><br><span class="line">        <span class="keyword">if</span> seq[<span class="number">4</span>:][i] == a[<span class="number">1</span>] + a[<span class="number">1</span>]:</span><br><span class="line">            genotypes[i] = np.array([<span class="number">1</span>,<span class="number">1</span>])</span><br><span class="line">    genotypes = genotypes.astype(<span class="string">'int'</span>).T</span><br><span class="line">    <span class="keyword">return</span> genotypes</span><br></pre></td></tr></table></figure>
<p>真实转换时python的转换速度较慢，这里为了展示基本的原理使用了python来演示。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">snps = snps.astype(<span class="string">'str'</span>)</span><br><span class="line">snps.shape</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">convert_2bit(snps[<span class="number">2</span>].astype(<span class="string">'str'</span>))</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">geno_conv = np.ndarray([<span class="number">10000</span>,<span class="number">6210</span>])</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> tqdm(range(<span class="number">5000</span>)):</span><br><span class="line">    geno_conv[i*<span class="number">2</span>:(i+<span class="number">1</span>)*<span class="number">2</span>] = convert_2bit(snps[i+<span class="number">1</span>])</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#查看SNP的大致情况，取前100行（50个位点），前50个样本，</span></span><br><span class="line">fig, ax = plt.subplots(figsize=(<span class="number">20</span>,<span class="number">10</span>))</span><br><span class="line">ax.matshow(geno_conv[:<span class="number">100</span>,:<span class="number">50</span>],cmap = cm.binary_r)</span><br></pre></td></tr></table></figure>
<p><img src="http://i1.fuimg.com/640680/870c6a096a867529.png" alt="Markdown"></p>
<h3 id="模型鲁棒性测试"><a href="#模型鲁棒性测试" class="headerlink" title="模型鲁棒性测试"></a>模型鲁棒性测试</h3><h4 id="设计特殊的交叉验证方式"><a href="#设计特殊的交叉验证方式" class="headerlink" title="设计特殊的交叉验证方式"></a>设计特殊的交叉验证方式</h4><p>不同的样本具有关联性，有的样本可能来自同一亲本，而训练集和测试集的划分并不是随机的<br>因此我们可以专门设计交叉验证时数据集的划分方式：</p>
<ul>
<li>random</li>
<li>by female</li>
<li>by male</li>
<li>cross sampling</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> method <span class="keyword">in</span> (<span class="string">'random'</span>, <span class="string">'by_female'</span>, <span class="string">'by_male'</span>, <span class="string">'cross'</span>):</span><br><span class="line">    <span class="keyword">with</span> h5py.File(<span class="string">'data/cv_index.%s'</span>%method, <span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">        index_train = f[<span class="string">'0/train'</span>][:]</span><br><span class="line">        index_test = f[<span class="string">'0/test'</span>][:]</span><br><span class="line">    fig, ax = subplots(<span class="number">2</span>, <span class="number">1</span>, figsize=(<span class="number">16</span>, <span class="number">6</span>))</span><br><span class="line">    sampling_table = np.zeros(np.prod(parent_table.shape))</span><br><span class="line">    sampling_table[index_train] = <span class="number">1</span></span><br><span class="line">    sampling_table = np.take(sampling_table, parent_table)</span><br><span class="line">    ax[<span class="number">0</span>].imshow(sampling_table, cmap=cm.Greys)</span><br><span class="line">    ax[<span class="number">0</span>].set_title(<span class="string">'Training samples (%s)'</span>%method)</span><br><span class="line"></span><br><span class="line">    sampling_table = np.zeros(np.prod(parent_table.shape))</span><br><span class="line">    sampling_table[index_test] = <span class="number">1</span></span><br><span class="line">    sampling_table = np.take(sampling_table, parent_table)</span><br><span class="line">    ax[<span class="number">1</span>].imshow(sampling_table, cmap=cm.Greys)</span><br><span class="line">    ax[<span class="number">1</span>].set_title(<span class="string">'Test samples (%s)'</span>%method)</span><br><span class="line">    plt.tight_layout()</span><br></pre></td></tr></table></figure>
<p>请在jupyter文件中查看相关图片。</p>
<h3 id="ANOVA加速算法"><a href="#ANOVA加速算法" class="headerlink" title="ANOVA加速算法"></a>ANOVA加速算法</h3><p>思路简要提示：</p>
<p><img src="http://i1.fuimg.com/640680/ef4e5c183a271656.png" alt="Markdown"></p>

      
    </div>
    
    
    

    

    <div>
    
    <div>
    
        <div style="text-align:center;color: #ccc;font-size:14px;">-----The ---- end ----<i class="fa fa-paw"></i>--- Thanks --- for --- Reading----</div>
    
</div>

    
    </div>

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/blog/tags/machine-learning/" rel="tag"><i class="fa fa-tag"></i> machine learning</a>
          
            <a href="/blog/tags/python/" rel="tag"><i class="fa fa-tag"></i> python</a>
          
            <a href="/blog/tags/bioinformatics/" rel="tag"><i class="fa fa-tag"></i> bioinformatics</a>
          
            <a href="/blog/tags/teaching/" rel="tag"><i class="fa fa-tag"></i> teaching</a>
          
            <a href="/blog/tags/TA/" rel="tag"><i class="fa fa-tag"></i> TA</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/blog/2018/10/06/55_python_basics/" rel="next" title="Python Tutorial">
                <i class="fa fa-chevron-left"></i> Python Tutorial
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/blog/2018/10/09/94-signal-project/" rel="prev" title="MDN for Signal Position Prediction">
                MDN for Signal Position Prediction <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/blog/images/avatar.png"
                alt="James Chen" />
            
              <p class="site-author-name" itemprop="name">James Chen</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/blog/archives/">
              
                  <span class="site-state-item-count">100</span>
                  <span class="site-state-item-name">posts</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/blog/categories/index.html">
                  <span class="site-state-item-count">10</span>
                  <span class="site-state-item-name">categories</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/blog/tags/index.html">
                  <span class="site-state-item-count">92</span>
                  <span class="site-state-item-name">tags</span>
                </a>
              </div>
            

          </nav>

          
            <div class="feed-link motion-element">
              <a href="/blog/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
                
                  <span class="links-of-author-item">
                    <a href="https://github.com/james20141606" target="_blank" title="GitHub">
                      
                        <i class="fa fa-fw fa-github"></i>GitHub</a>
                  </span>
                
                  <span class="links-of-author-item">
                    <a href="mailto:xp-chen14@mails.tsinghua.edu.cn" target="_blank" title="E-Mail">
                      
                        <i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  </span>
                
            </div>
          

          
          

          
          

          <script type='text/javascript' id='clustrmaps' src='//cdn.clustrmaps.com/map_v2.js?cl=d9dbc8&w=a&t=n&d=MGAfMNv-Snrv0Yg1d2t2EH3ATBUCJtdgbna_qHvb4AI&co=302dad'></script>

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#eMaize玉米育种挑战赛"><span class="nav-number">1.</span> <span class="nav-text">eMaize玉米育种挑战赛</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#eMaize背景简介"><span class="nav-number">1.1.</span> <span class="nav-text">eMaize背景简介</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#编程工具介绍"><span class="nav-number">1.2.</span> <span class="nav-text">编程工具介绍</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#数据介绍"><span class="nav-number">1.3.</span> <span class="nav-text">数据介绍</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Genotype数据"><span class="nav-number">1.3.1.</span> <span class="nav-text">Genotype数据</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#SNP数据存储格式"><span class="nav-number">1.3.1.1.</span> <span class="nav-text">SNP数据存储格式</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#性状数据"><span class="nav-number">1.3.1.2.</span> <span class="nav-text">性状数据</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#查看性状的分布情况"><span class="nav-number">1.3.1.2.1.</span> <span class="nav-text">查看性状的分布情况</span></a></li></ol></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Quiz具体要求"><span class="nav-number">1.4.</span> <span class="nav-text">Quiz具体要求</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#特征选择和冗余特征筛除"><span class="nav-number">1.4.1.</span> <span class="nav-text">特征选择和冗余特征筛除</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#完成对测试集三种性状的预测，尝试得到尽可能好的预测结果"><span class="nav-number">1.4.2.</span> <span class="nav-text">完成对测试集三种性状的预测，尝试得到尽可能好的预测结果</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#加分内容"><span class="nav-number">1.4.3.</span> <span class="nav-text">加分内容</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#对模型进行鲁棒性测试-10’"><span class="nav-number">1.4.3.1.</span> <span class="nav-text">对模型进行鲁棒性测试 (10’)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#使用ANOVA进行特征选择的加速算法-5’"><span class="nav-number">1.4.3.2.</span> <span class="nav-text">使用ANOVA进行特征选择的加速算法 (5’)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#混合线性模型-20’"><span class="nav-number">1.4.3.3.</span> <span class="nav-text">混合线性模型 (20’)</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#补充知识（选读）"><span class="nav-number">1.5.</span> <span class="nav-text">补充知识（选读）</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#测试预测结果"><span class="nav-number">1.5.1.</span> <span class="nav-text">测试预测结果</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#查看训练集与测试集的划分"><span class="nav-number">1.5.1.0.1.</span> <span class="nav-text">查看训练集与测试集的划分</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#测试集具体划分"><span class="nav-number">1.5.1.0.2.</span> <span class="nav-text">测试集具体划分</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#评价指标"><span class="nav-number">1.5.1.0.3.</span> <span class="nav-text">评价指标</span></a></li></ol></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#将SNP数据编码为向量"><span class="nav-number">1.5.2.</span> <span class="nav-text">将SNP数据编码为向量</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#具体处理过程"><span class="nav-number">1.5.2.1.</span> <span class="nav-text">具体处理过程</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#模型鲁棒性测试"><span class="nav-number">1.5.3.</span> <span class="nav-text">模型鲁棒性测试</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#设计特殊的交叉验证方式"><span class="nav-number">1.5.3.1.</span> <span class="nav-text">设计特殊的交叉验证方式</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#ANOVA加速算法"><span class="nav-number">1.5.4.</span> <span class="nav-text">ANOVA加速算法</span></a></li></ol></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        
<div class="copyright">&copy; <span itemprop="copyrightYear">2021</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">James Chen</span>

  
</div>



  <span class="post-meta-divider">|</span>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共207.5k字</span>
</div>


        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  
    <script type="text/javascript" src="/blog/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/blog/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  



  


  <script type="text/javascript" src="/blog/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/blog/js/src/motion.js?v=5.1.4"></script>



  
  

  
  <script type="text/javascript" src="/blog/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/blog/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/blog/js/src/bootstrap.js?v=5.1.4"></script>



  


  

    
      <script id="dsq-count-scr" src="https://cmwonderland.disqus.com/count.js" async></script>
    

    
      <script type="text/javascript">
        var disqus_config = function () {
          this.page.url = 'https://www.cmwonderland.com/blog/2018/10/06/31_quiz_emaize_tutorial/';
          this.page.identifier = '2018/10/06/31_quiz_emaize_tutorial/';
          this.page.title = 'QUIZ Predicting maize traits using SNP data';
        };
        var d = document, s = d.createElement('script');
        s.src = 'https://cmwonderland.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      </script>
    

  




	





  














  





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.4.js"></script>
  <script>AV.initialize("klOjl0RBA8qP5IKgIXkOszBr-gzGzoHsz", "rCaN5wX4mjiRkRMzP95g7XHz");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  

  
<script>
(function(){
    var bp = document.createElement('script');
    var curProtocol = window.location.protocol.split(':')[0];
    if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
    }
    else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
    }
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
  


  

  

  
</body>
</html>
